src/test/java/com/google/api/codegen/testdata/csharp_gapic_resourcenames_library.baseline

```diff
@@ -0,0 +1,719 @@
+            _nameToString = GaxPreconditions.CheckNotNull(nameToString, nameof(nameToString));
+            _stringToName = GaxPreconditions.CheckNotNull(stringToName, nameof(stringToName));
+        }
+
+        private IList<string> _underlyingList;
+        private Func<TResource, string> _nameToString;
+        private Func<string, TResource> _stringToName;
+
+        public int Count => _underlyingList.Count;
+        public bool IsReadOnly => _underlyingList.IsReadOnly;
+        public TResource this[int i]
+        {
+            get { return _stringToName(_underlyingList[i]); }
+            set { _underlyingList[i] = _nameToString(value); }
+        }
+        public void Add(TResource value) => _underlyingList.Add(_nameToString(value));
+        public void Clear() => _underlyingList.Clear();
+        public bool Contains(TResource value) => _underlyingList.Contains(_nameToString(value));
+        public void CopyTo(TResource[] array, int arrayIndex)
+        {
+            // TODO: Improve exceptions
+            if (array == null) throw new ArgumentNullException();
+            if (arrayIndex < 0) throw new ArgumentOutOfRangeException();
+            if (Count > array.Length - arrayIndex) throw new ArgumentException();
+            for (int i = 0; i < Count; i++)
+            {
+                array[arrayIndex + i] = _stringToName(_underlyingList[i]);
+            }
+        }
+        public IEnumerator<TResource> GetEnumerator() => _underlyingList.Select(_stringToName).GetEnumerator();
+        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() => GetEnumerator();
+        public int IndexOf(TResource value) => _underlyingList.IndexOf(_nameToString(value));
+        public void Insert(int index, TResource item) => _underlyingList.Insert(index, _nameToString(item));
+        public bool Remove(TResource item) => _underlyingList.Remove(_nameToString(item));
+        public void RemoveAt(int index) => _underlyingList.RemoveAt(index);
+    }
+
+    /// <summary>
+    /// Resource name for a shelf.
+    /// </summary>
+    public sealed partial class ShelfName : ISingleResourceName
+    {
+        private static readonly PathTemplate s_template = new PathTemplate("shelves/{shelf_id}");
+
+        /// <summary>
+        /// Parses a shelf resource name in string form into a <see cref="ShelfName"/>.
+        /// </summary>
+        /// <param name="shelfName">The shelf resource name in string form. Must not be <c>null</c>.</param>
+        public static ShelfName Parse(string shelfName)
+        {
+            GaxPreconditions.CheckNotNull(shelfName, nameof(shelfName));
+            ResourceName resourceName = s_template.ParseName(shelfName);
+            return new ShelfName(resourceName[0]);
+        }
+
+        public static bool TryParse(string shelfName, out ShelfName result)
+        {
+            GaxPreconditions.CheckNotNull(shelfName, nameof(shelfName));
```

Comment:
All the redundancy here feels annoying. I suspect we can get this down to

```csharp
public static bool TryParse(string shelfName, out ShelfName result) =>
    s_template.TryParseName(shelfName, nameof(shelfName), out result, p0 => new ShelfName(p0));
```

... and likewise for up to (say) 4 parameters. Can definitely do that later though.

Ditto for `Parse`:

```csharp
public static ShelfName Parse(string shelfName) =>
    s_template.ParseName(shelfName, nameof(shelfName), p0 => new ShelfName(p0));
```


(With GAX changes, of course.)